### ISSUE TITLE
5 gigabytes of session lock files

### ISSUE THREAD

**Issue by ampanasiuk** [cherrypy/cherrypy#1855]
**I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

**Do you want to request a *feature* or report a *bug*?**

**What is the current behavior?**
I recently found out hundreds of thousands of session lock files totalling 5GB. How does this happen?

I have a "status" page handler that returns system status -- CPU use percent, whether a service is running or not, etc. The web app requests this URL once a second and shows the result in a bar at the top of the page.

If you keep the browser tab on the web interface, then even after the session expires the web page doesn't reload or anything and the status keeps on being displayed. But with the session and the cookie in the client expired, the status URL is requested with no cookie. Cherrypy creates a new session id and session lock file. So the next second the browser uses the new session id. However, as nothing was stored in the session from the status page handler, the session file never materialised -- so cherrypy creates a new session id, a new session lock file and this keeps happening for each subsequent second.

This is a regression after #1779, FileSession's lock file is no longer deleted.

Workaround is to set explicit session locking for this handler, but that's just a workaround, not a fix.

**What is the expected behavior?**

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

- Cheroot version: X.X.X
- CherryPy version: 18.5.0
- Python version: 3.6.X
- OS: XXX
- Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

**Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, e.g. stackoverflow, gitter, etc.)

**Comment by ikus060** [comment]
Bumping this issue. I'm having this problem too with CherryPy v18.6.0

Session files get properly cleanup, but the lock files are not.

**Comment by ikus060** [comment]
As an implementation recommendation, I would simply add a clean-up process to delete the file if the modification date is greater then the expiry. That way you may keep the lock file around, but still get them clean-up.

**Comment by webknjaz** [comment]
Yeah, it's a known issue. The cleanup is not that easy to get right in a cross-platform manner, unfortunately. IIRC there were at least problems with Windows locking the access to files with some race conditions around that. But a PR is welcome :)

**Comment by ikus060** [comment]
To support older cherrypy version, I intend to get it fixed outside Cherrypy for now with a clean-up cronjob.

**Comment by ampanasiuk** [comment]
Another work around is to write or read something from the session.
```python3
_ = cherrypy.session.get("this line is a work around for https://github.com/cherrypy/cherrypy/issues/1855")
```

**Comment by jiangwen365** [comment]
> Yeah, it's a known issue. The cleanup is not that easy to get right in a cross-platform manner, unfortunately. IIRC there were at least problems with Windows locking the access to files with some race conditions around that. But a PR is welcome :)

I also see this problem happening when I ported my site's source code from Python 2 to 3. However, it looks when CherryPy is running on Python 2 this problem does not exist or it is much better.

**Comment by jiangwen365** [comment]
> > Yeah, it's a known issue. The cleanup is not that easy to get right in a cross-platform manner, unfortunately. IIRC there were at least problems with Windows locking the access to files with some race conditions around that. But a PR is welcome :)
> 
> I also see this problem happening when I ported my site's source code from Python 2 to 3. However, it looks when CherryPy is running on Python 2 this problem does not exist or it is much better.

Today I found out this is not related to Python 2/3. 
It seems if I downgrade CherryPy from 18.6.1 to CherrPy 17.4.2 (on Python 3.9), all these lock files would not be created at all.

**Comment by jiangwen365** [comment]
It might be related to this commit: 2543184468b03d8ba8bbc39da35a882bf76e4da4

**Comment by jiangwen365** [comment]
Never mind I realize the core development team already know this: https://github.com/cherrypy/cherrypy/issues/1779

**Comment by jiangwen365** [comment]
For those who might be interested, I now regularly run the below python code to clean session lock files. The code compares lock files with existing session files and removes those that don't have matching session files.

```python
SESSION_PREFIX = 'session-'
LOCK_SUFFIX = '.lock'

all_files = {fname for fname in os.listdir(storage_path) if fname.startswith(SESSION_PREFIX)}
lock_files = {fname for fname in all_files if fname.endswith(LOCK_SUFFIX)}
session_files = all_files - lock_files
session_lock_files = {fname + LOCK_SUFFIX for fname in session_files}


for fname in lock_files - session_lock_files:
    try:
        os.unlink(os.path.join(storage_path, fname))
    except OSError:
        pass
```
